### DISCUSSION TITLE
Conflict - Duplicate primary key on Push

### DISCUSSION THREAD

**Discussion by henda79** [Azure/azure-mobile-apps#418] [Q&A] [answered]
I've implemented the framework into my code after doing some testing with the Todo demo, and I'm getting an exception when pushing my table with a single row. I get a conflict telling me the Primary Key is a duplicate.

Is it supposed to push the same row every push, even though it already exists at the server ?
and oddly I get the error even if I stop the local server service.

I'm fairly sure I've setup correctly, my client models inherit `DatasyncClientData` and the server models `EntityTableData`
I was using version 5.0.4 but just updated to 5.0.6 and its the same.

Any pointers ?

EDIT: so after tracing the code in the SDK, I see the client stores errors in the db and this will explain why I get this exception without a server, but how did it happen with only a single row and a single insert. Just to clarify the row I inserted on the client has appeared on the server and looks fine.

**Comment by adrianhall** [comment]
You would get that sort of conflict by:

a) Doing a CreateAsync on the client and inserting the same data in the server
b) Using an IdGenerator which doesn't provide globally unique IDs

If Push cannot create the item, you have to do something with the queued item - cancel and update, cancel and discard - you can't just expect it to silently ignore the item.

**Reply by henda79** [reply]
On the client I use
`
_client.GetOfflineTable<T>().InsertItemAsync(item);
`

item is the model inherited from `DatasyncClientData`, then I do

`
_client.GetOfflineTable<T>().PushItemsAsync();
`

with a fresh client db, the push works and creates at the server, but then if I try to 'sync' by push/pull on the table, I get the exception.

I don't use IdGenerator at the moment.

Is it Ok, to use `GetOfflineTable` like that ? as in with each call to insert / update etc.

**Comment by adrianhall** [comment]
Did you do a pull afterwards?

Generally, you want to make the table a singleton.   The client **has** to be a singleton (it holds the SyncContext, which also has to be a singleton because that's what holds onto the database connection).

**Comment by henda79** [comment]
So I have spent some time and gone over the code and have fixed the issue of the duplicate key, but now after sync I end up with a duplicate on the client, no exceptions though.

I can work offline and create the record on the client and its is not duplicated, but after a sync, the server contains 1 record but the client has 2 but oddly they seem to have the same Id.

**Reply by adrianhall** [reply]
the Id is a key field - SQLite doesn't allow you to have duplicates. This points to something very basic going awry.  

Without a full code, we can't diagnose this.

**Reply by henda79** [reply]
Fixed, seems like the issue was coming from the converting the string Id to Guid on the client side.

Do you know of a way to easily ready the tables and data in the client Sqlite Db ? If tried various tool but they can view files in other apps space which I know I is normal. I have tried to relocate the Db to a readable folder such as Downloads but can't seem to get the file URI correct.

Android Device Monitor no longer works when targeting Android 12.

**Reply by adrianhall** [reply]
1) Output the connection string using a debug statement.
2) Copy the .db file from the device to your workstation.
3) There are SQLite tools in VS, VSCode, and standalone for looking at it.

**Reply by henda79** [reply]
I've can see where the Db is saved to but only the app can access it unless the device is rooted (which its not).

Do you know what the connection string would need to be to save the file in the Downloads folder, as I've tried many and nothing works, just keep getting an exception stating the connection string is invalid.

I'm using an emulator Btw.

**Comment by henda79** [comment, marked as answer]
I 've worked out what is going on. It had to do with the way the Id's are stored in Sqlite. Some times the Id was stored with or without hyphens in the Guid string.
